### ISSUE TITLE
Bug: [@typescript-eslint/no-unnecessary-condition] false positive

### ISSUE THREAD

**Issue by Maxim-Mazurok** [typescript-eslint/typescript-eslint#9477]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.ts&code=G4ewlgJgBAFAhgZwJ4DsDGsCUUC8A%2BKAbwFgAoKKNEFBAFyhACcwBzMFOAGwGFqEROAUwAyIFrkp8BggHScxAbjIUh9JsC4AxdmAQALQRG0pBEgGZcEgpaWWSa9EwHdeNaaJYAuKLSQAHQRAze34hOTEJGBlovzhGOABbQVpBRgRsfCI7CjBgmF8AoKhY%2BKSUtIBtAAYAXVwcHCgAIjpmFBYmqAAyLuK4xOTUhGqamXY0TgBXCEEEGCbGQTgIJB8QKEmrJsxsEnIKA4B6Q6hBAA8AtHomwB4NwEj9qDg-MCgAWigATRBJxigAeQCKAAggAFACSUAQlygekQUAARoJBCh7MBUiloOxaOtFstVtiNlYGIwNJwAIRNbIHdRaHT6QzGUyNWiMSbWKkAXypuVgBUCwRKA3Kw1q9UaLRZ7A63V6grKQxGMmxoicqW4iEEMEwY3QUxmcyaqUYTG2uypFGOpwugiuzUAvBuARb2KE8Xu8ACr%2BQQAUUYxsY3nVKBQIHouOgfmNAUYtDAswYwUmKBmZnYhlgYalUAA5BGQFGY7Ms5hKfsDsTScZdAYjKnzJZ2aWoFzbI2mKx2FxXKERGIZE8-JwkDAqG4hAAaPqlQZpTA2CgcuchaThcSNZxd9yKMh2Ed0ctcCRwJxwMD0MAJPxMWjzGmcbYKKCW8%2BX6OPMzlJdheTiXRQRYDuA0EMOwjxPNQSS4GQWGRVI4BSLUH0tXdl2-KBf0RTM0C4IRMRoSBTCKWg9F-KgZm3fYeRgMlb0rekaxMXZHxOAAeZi0IQBA2XYmFUkECdT244N6C4Y8kAQHxWSIpAqVzICOJkc5TxgABGWc7GbDlMAQsggA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
void (async () => {
  const originalConsoleLog = console.log;
  let orvalFinishedFine = false;

  const newConsoleLog: typeof console.log = (...parameters) => {
    if (typeof parameters[0] === "string" && parameters[0].includes("ready to use")) {
      // expect "🎉 api - Your OpenAPI spec has been converted into ready to use orval!"
      orvalFinishedFine = true;
    }
    if (typeof parameters[0] === "string" && parameters[0].toLowerCase().includes("error")) {
      // expect "🛑  api - TypeError: Cannot read properties of undefined (reading 'properties')"
      orvalFinishedFine = false;
    }

    originalConsoleLog.apply(console, parameters);
  };
  console.log = newConsoleLog;

  const orval = await import("orval"); // import after console.log is replaced
  await orval.generate(); // console.log is being called inside of this code

  if (!orvalFinishedFine) { // << issue is here, it is not always truthy
    process.exit(1);
  }
})();
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No false positive

### Actual Result

false positive

**Comment by bradzacher** [comment]
The rule is powered by types.

According to TS the variable is typed as `false`.
You can verify this by hovering over the variable name.

TS does not analyse the side-effects of function calls so it cannot understand that your function mutates the variable.

**Comment by Josh-Cena** [comment]
This is microsoft/TypeScript#9998; the simplest fix is `let orvalFinishedFine = false as boolean;`
